File: kafka/kafka-annotation-processor/src/main/java/ru/tinkoff/kora/kafka/annotation/processor/KafkaConsumerGenerator.java
package ru.tinkoff.kora.kafka.annotation.processor;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.common.Tag;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

import static ru.tinkoff.kora.kafka.annotation.processor.KafkaClassNames.*;
import static ru.tinkoff.kora.kafka.annotation.processor.utils.KafkaUtils.prepareMethodName;
import static ru.tinkoff.kora.kafka.annotation.processor.utils.KafkaUtils.prepareTagName;

public class KafkaConsumerGenerator {
    private final ProcessingEnvironment processingEnv;
    private final Types types;
    private final Elements elements;
    private final TypeMirror exceptionType;
    private final TypeMirror consumerRecordType;
    private final TypeMirror consumerRecordsType;
    private final TypeMirror consumerType;

    public KafkaConsumerGenerator(ProcessingEnvironment processingEnv) {
        this.processingEnv = processingEnv;
        this.elements = processingEnv.getElementUtils();
        this.types = processingEnv.getTypeUtils();

        this.exceptionType = this.elements.getTypeElement(Exception.class.getCanonicalName()).asType();
        this.consumerRecordType = this.types.erasure(this.elements.getTypeElement(KafkaClassNames.consumerRecord.canonicalName()).asType());
        this.consumerRecordsType = this.types.erasure(this.elements.getTypeElement(KafkaClassNames.consumerRecords.canonicalName()).asType());
        this.consumerType = this.types.erasure(this.elements.getTypeElement(KafkaClassNames.consumer.canonicalName()).asType());
    }

    @Nullable
    public MethodSpec generate(ExecutableElement executableElement) {
        var listenerAnnotation = AnnotationUtils.findAnnotation(executableElement, kafkaIncoming);
        if (listenerAnnotation == null) {
            return null;
        }
        var controller = (TypeElement) executableElement.getEnclosingElement();
        var methodName = prepareMethodName(controller.getSimpleName().toString(), executableElement.getSimpleName().toString());

        var methodBuilder = MethodSpec.methodBuilder(methodName)
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .addParameter(TypeName.get(controller.asType()), "_controller");

        var tagName = prepareTagName(controller.getSimpleName().toString(), executableElement.getSimpleName().toString());
        var tagsBlock = CodeBlock.builder().add("{");
        tagsBlock.add("$L.class", tagName);
        tagsBlock.add("}");

        var propertiesParameter = ParameterSpec
            .builder(kafkaConsumerConfig, "_consumerConfig");

        propertiesParameter.addAnnotation(
            AnnotationSpec.builder(Tag.class).addMember("value", tagsBlock.build()).build()
        );

        methodBuilder.addParameter(propertiesParameter.build());
        var consumerData = extractConsumerData(executableElement);

        if (consumerData == null) {
            throw new ProcessingErrorException("Unsupported signature for @KafkaIncoming", executableElement);
        }

        methodBuilder.addParameter(ParameterizedTypeName.get(deserializer, consumerData.keyType()), "keyDeserializer");
        methodBuilder.addParameter(ParameterizedTypeName.get(deserializer, consumerData.valueType()), "valueDeserializer");
        methodBuilder.addParameter(ParameterizedTypeName.get(kafkaConsumerTelemetry, consumerData.keyType(), consumerData.valueType()), "telemetry");
        methodBuilder.returns(ParameterizedTypeName.get(kafkaConsumerContainer, consumerData.keyType(), consumerData.valueType()));

        var returnBlock = CodeBlock.of(
            """
                    return new $T<>(
                            _consumerConfig,
                            keyDeserializer,
                            valueDeserializer,
                            $T.wrapHandler(telemetry, _controller::$L)
                );""",
            kafkaConsumerContainer,
            handlerWrapper,
            executableElement.getSimpleName().toString()
        );
        methodBuilder.addCode(returnBlock);

        return methodBuilder.build();
    }

    private ConsumerContainerData extractConsumerData(ExecutableElement executableElement) {
        var params = executableElement.getParameters();

        if (params.size() == 3) {

            if (!this.types.isSameType(params.get(2).asType(), this.exceptionType)) {
                return null;
            }

            return new ConsumerContainerData(
                TypeName.get(params.get(0).asType()),
                TypeName.get(params.get(1).asType())
            );

        } else if (params.size() == 2) {
            TypeMirror erasure = this.types.erasure(params.get(0).asType());
            if (this.types.isSameType(erasure, consumerRecordsType) || this.types.isSameType(erasure, consumerRecordType)) {
                if (!this.types.isSameType(this.types.erasure(params.get(1).asType()), consumerType)) {
                    throw new ProcessingErrorException("Second argument should have type org.apache.kafka.clients.consumer.Consumer", executableElement);
                }

                var args = ((DeclaredType) params.get(0).asType()).getTypeArguments();
                return new ConsumerContainerData(
                    TypeName.get(args.get(0)),
                    TypeName.get(args.get(1))
                );
            }

            if (this.types.isSameType(params.get(1).asType(), this.exceptionType)) {
                return new ConsumerContainerData(
                    TypeName.get(byte[].class),
                    TypeName.get(params.get(0).asType())
                );
            }

            return new ConsumerContainerData(
                TypeName.get(params.get(0).asType()),
                TypeName.get(params.get(1).asType())
            );

        } else if (params.size() == 1) {
            TypeMirror erasure = this.types.erasure(params.get(0).asType());
            if (this.types.isSameType(erasure, consumerRecordsType) || this.types.isSameType(erasure, consumerRecordType)) {
                var args = ((DeclaredType) params.get(0).asType()).getTypeArguments();
                return new ConsumerContainerData(
                    TypeName.get(args.get(0)),
                    TypeName.get(args.get(1))
                );
            }

            return new ConsumerContainerData(
                TypeName.get(byte[].class),
                TypeName.get(params.get(0).asType())
            );
        }

        return null;
    }


}

File: kafka/kafka-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kafka/symbol/processor/KafkaConsumerGenerator.kt
package ru.tinkoff.kora.kafka.symbol.processor

import com.google.devtools.ksp.*
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.Modifier
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.ConsumerRecords
import org.apache.kafka.common.serialization.Deserializer
import ru.tinkoff.kora.common.Tag
import ru.tinkoff.kora.kafka.common.annotation.KafkaIncoming
import ru.tinkoff.kora.kafka.common.config.KafkaConsumerConfig
import ru.tinkoff.kora.kafka.common.containers.KafkaConsumerContainer
import ru.tinkoff.kora.kafka.common.containers.handlers.wrapper.HandlerWrapper
import ru.tinkoff.kora.kafka.common.telemetry.KafkaConsumerTelemetry
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException

@KspExperimental
class KafkaConsumerGenerator(private val kspLogger: KSPLogger, resolver: Resolver) {
    private val nullableExceptionType = resolver.getClassDeclarationByName(java.lang.Exception::class.qualifiedName!!)!!.asType(emptyList()).makeNullable()
    private val consumerType = resolver.getClassDeclarationByName(Consumer::class.qualifiedName!!)!!.asStarProjectedType()
    private val consumerRecordType = resolver.getClassDeclarationByName(ConsumerRecord::class.qualifiedName!!)!!.asStarProjectedType()
    private val consumerRecordsType = resolver.getClassDeclarationByName(ConsumerRecords::class.qualifiedName!!)!!.asStarProjectedType()

    fun generate(functionDeclaration: KSFunctionDeclaration): FunSpec? {
        functionDeclaration.getAnnotationsByType(KafkaIncoming::class).firstOrNull() ?: return null
        val controller = functionDeclaration.parentDeclaration as KSClassDeclaration
        val methodName = prepareMethodName(controller.simpleName.asString(), functionDeclaration.simpleName.asString())
        val funBuilder = FunSpec.builder(methodName)
            .addParameter("_controller", controller.toClassName())
        val tagName = prepareTagName(controller.simpleName.asString(), functionDeclaration.simpleName.asString())
        val tagsBlock = CodeBlock.builder().add("%L::class", tagName)
        val propertiesParameter = ParameterSpec.builder("_consumerConfig", KafkaConsumerConfig::class)
        propertiesParameter.addAnnotation(
            AnnotationSpec.builder(Tag::class).addMember(tagsBlock.build()).build()
        )
        funBuilder.addParameter(propertiesParameter.build())
        val consumerData = extractConsumerData(functionDeclaration)

        funBuilder.addParameter(
            "keyDeserializer", Deserializer::class.asClassName().parameterizedBy(consumerData.keyType)
        )
        funBuilder.addParameter(
            "valueDeserializer", Deserializer::class.asClassName().parameterizedBy(consumerData.valueType)
        )
        funBuilder.addParameter(
            "telemetry", KafkaConsumerTelemetry::class.asClassName().parameterizedBy(consumerData.keyType, consumerData.valueType)
        )
        funBuilder.returns(
            KafkaConsumerContainer::class.asClassName().parameterizedBy(consumerData.keyType, consumerData.valueType)
        )
        val returnBlock = CodeBlock.of(
            """
                    return %T(
                            _consumerConfig,
                            keyDeserializer,
                            valueDeserializer,
                            %T.wrapHandler(telemetry, _controller::%L)
                )""",
            KafkaConsumerContainer::class,
            HandlerWrapper::class,
            functionDeclaration.simpleName.asString()
        )
        funBuilder.addCode(returnBlock)
        return funBuilder.build()
    }

    private fun extractConsumerData(ksFunctionDeclaration: KSFunctionDeclaration): ConsumerContainerData {
        val params = ksFunctionDeclaration.parameters
        val isSuspend = ksFunctionDeclaration.modifiers.contains(Modifier.SUSPEND)
        if (isSuspend) {
            throw ProcessingErrorException("Unsupported signature for @KafkaIncoming: suspend function", ksFunctionDeclaration)
        }

        when(params.size) {
            3 -> {
                if (params[2].type.resolve() != nullableExceptionType) {
                    throw ProcessingErrorException("Unsupported signature for @KafkaIncoming: 3rd argument should be 'Exception?'", ksFunctionDeclaration)
                }

                return ConsumerContainerData(
                    params[0].type.toTypeName(),
                    params[1].type.toTypeName()
                )
            }

            2 -> {
                val firstParamType = params[0].type.resolve()
                val secondParamType = params[1].type.resolve()

                val firstParamTypeProjection = params[0].type.resolve().starProjection()
                if ((firstParamTypeProjection == consumerRecordType || firstParamTypeProjection == consumerRecordsType)) {
                    if (secondParamType.starProjection() != consumerType) {
                        throw ProcessingErrorException("Second argument should have type org.apache.kafka.clients.consumer.Consumer", ksFunctionDeclaration);
                    }

                    return ConsumerContainerData(
                        firstParamType.arguments[0].toTypeName(),
                        firstParamType.arguments[1].toTypeName()
                    )
                }

                if (secondParamType == nullableExceptionType) {
                    return ConsumerContainerData(
                        ByteArray::class.asTypeName(),
                        params[0].type.toTypeName()
                    )
                }


                return ConsumerContainerData(
                    params[0].type.toTypeName(),
                    params[1].type.toTypeName()
                )
            }

            1 -> {
                val firstParamType = params[0].type.resolve()
                val firstParamTypeProjection = params[0].type.resolve().starProjection()

                if ((firstParamTypeProjection == consumerRecordType || firstParamTypeProjection == consumerRecordsType)) {
                    return ConsumerContainerData(
                        firstParamType.arguments[0].toTypeName(),
                        firstParamType.arguments[1].toTypeName()
                    )
                }

                return ConsumerContainerData(
                    ByteArray::class.asTypeName(),
                    params[0].type.toTypeName()
                )
            }
        }

        throw ProcessingErrorException(ProcessingError("Unsupported signature for @KafkaIncoming", ksFunctionDeclaration))
    }

}
